Add accessibility statement

isabela-pf opened the pull request:
This is a draft accessibility statement edited from the [W3C Accessibility Statement Generator Tool](https://www.w3.org/WAI/planning/statements/). It only applies to JupyterLab.

---

Below are thoughts copied from [quansight-labs/jupyter-a11y-mgmt #52](https://github.com/Quansight-Labs/jupyter-a11y-mgmt/issues/52#issuecomment-1127886714).

## Things I think we should add

From [Make your website or app accessible and publish an accessibility statement  - Gov.UK](https://www.gov.uk/guidance/make-your-website-or-app-accessible-and-publish-an-accessibility-statement), there are a few things we need to do. I've added check boxes to the quote to show what I think is covered in the above draft.
> Your statement needs to cover:
> - [x] whether your website or mobile app is ‘fully’, ‘partially’ or ‘not’ compliant with accessibility standards
> - [x] if it’s not fully compliant, which parts of your website or mobile app do not currently meet accessibility standards and why (for example, because they are exempt or it would be a disproportionate burden to fix things)
> - [ ] how people can get alternatives to content that’s not accessible to them
> - [x] how to contact you to report accessibility problems - and a link to the website that they can use if they’re not happy with your response.

With this and the other resources, I think we may want to add
- How people can get alternatives to content that's not accessible to them (mentioned above).
- An [accessibility monitoring methodology](https://www.gov.uk/guidance/accessibility-monitoring-how-we-test) section (ie. the testing work). This could be a link to a document that goes more into detail about this process since I doubt it is the main focus for everyone reviewing this statement.
- Maybe a note about open source. We'd have to be sure not to let this sound like an excuse, but rather a note about response times and that there's no support for this work.
- A way for people to report they'd like to be reached out if/when we are testing (like [Gatsby's Surveys and Research section](https://www.gatsbyjs.com/accessibility-statement/#surveys-and-research-projects)). Since I think this could come off wrong, we may want to mention like "for outreach in any potential paid testing opportunities." At the same time, it'd be useful to know we have people available if we want to apply for funding this kind of work in the future.

## What next

To move forward, I would like to know a few things.

- [ ] What seems to be inaccurate or missing from this current draft?
- [ ] What's on your wish list of statement additions?
- [x] How would we like to collaborate on this next? Is it a good time to open a PR, or do y'all find it easier to iterate elsewhere first? Answer: iterating here is good.
- [x] I filled this out for JupyterLab. Is there any supporting project we think should be included (ex. the notebook file)? Answer: Not now. We're scoping it, but mentioning that JupyterLab does not exist in isolation.
- [ ] Where should this file go?

isabela-pf committed:
Create draft accessibility statement

trallard committed:
Merge branch 'master' into statement

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+* Include accessibility as part of our mission statement.
+* Provide continual accessibility training for our community.
+* Assign clear accessibility goals and responsibilities.
+* Employ formal accessibility quality assurance methods.
```
Idea: we should add something on documenting - changes, improvements and approaches to testing and QA

isabela-pf replied:
These were from a list of W3C suggested ones, so I'm unclear if they map to some kind of special requirements.

I added another point.

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+### Conformance status
+
+The <a href="https://www.w3.org/WAI/standards-guidelines/wcag/">Web Content Accessibility Guidelines (WCAG)</a> defines requirements for designers and developers to improve accessibility for people with disabilities. It defines three levels of conformance: Level A, Level AA, and Level AAA. JupyterLab is non conformant with WCAG 2.1 level AA. Non conformant means that the content does not conform  ...
```
Is this the auto-generated text? 

> Non conformant means that the content does not conform the
Using the word in the definition which makes it hard to follow/understand

trallard replied:
Stephanie Walter has an amazing illustration for this in https://stephaniewalter.design/blog/5-illustrations-to-understand-and-promote-accessibility/

Could we add this to add some visual sugar - of course with due acknowledgement and x-linking to the post

isabela-pf replied:
It is the auto-generated text; most of this is not my work alone.

I'll look into how I need to add the image with credit and add it.

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+### Feedback
+
+We welcome your feedback on the accessibility of JupyterLab. Please let us know if you encounter accessibility barriers on JupyterLab:
```
accessibility status? approach?

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+Known limitations for JupyterLab:
+
+1. **Documents**: Documents writen by the community may not include accessible content because we do not and cannot review every document that can be opened in JupyterLab. To support accessible documents, we are drafting guidelines for accessible document content with an emphasis on Jupyter notebooks. Please report the issue to the author and [open an issue on jupyter/accessibility](https://githu ...
```
Maybe add an example: "Documents such as notebooks, markdown files, etc."

trallard replied:
```suggestion
1. **Documents**: Documents written by the community may not include accessible content because we do not and cannot review every document that can be opened and edited in JupyterLab. To support accessible documents, we are drafting guidelines for accessible document content with an emphasis on Jupyter notebooks. Please report the issue to the author and [open an issue on jupyter/accessibility](https://github.com/jupyter/accessibility/issues/new) describing the problem and the behavior you expect so we may integrate it into our content guidelines.
```

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+1. **Documents**: Documents writen by the community may not include accessible content because we do not and cannot review every document that can be opened in JupyterLab. To support accessible documents, we are drafting guidelines for accessible document content with an emphasis on Jupyter notebooks. Please report the issue to the author and [open an issue on jupyter/accessibility](https://githu ...
+    
+2. **JupyterLab extensions**: JupyterLab extensions writen by the community may not be accessible because JupyterLab extensions can be written by anyone in the community and have no required review process. We do not and can not review every JupyterLab extension. To support accessible extensions, we encourage extension authors to use existing, accessible JupyterLab components for their extension. ...
```
```suggestion
2. **JupyterLab extensions**: JupyterLab extensions written by the community may not be accessible because JupyterLab extensions can be written by anyone in the community and have no standard review process. We do not and can not review every JupyterLab extension. To support accessible extensions, we encourage extension authors to use existing, accessible JupyterLab components for their extensions. We also provide periodic opportunities for community education on accessibility. Please report the issue to the author and let them know the [jupyter/accessibility](https://github.com/jupyter/accessibility/) community may be able to provide guidance.
```

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+* Self-evaluation
+* Regular automated testing to monitor for regressions (can be found at [jupyter/accessibility](https://github.com/jupyter/accessibility)
+* User feedback</li>
```
```suggestion
* User feedback
```
assuming that was left over

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+### Formal complaints
+
+* [Write an issue on jupyter/accessibility](https://github.com/jupyter/accessibility/issues/new)
```
Shall we create a template for this? or just rely on a blank issue?

isabela-pf replied:
I think adding a template could be useful, but I don't have one to add at the moment. Do you want me to hold off on this PR until that's done?

There's a list of templates that I believe are wanted for that repo, so I would probably do a few at once.

trallard replied:
Let's move templates to the next cycle and get this merged

trallard reviewed (commented):
Reviewed - added some comments and suggestions, nothing too major

trallard commented:
> * How people can get alternatives to content that's not accessible to them (mentioned above).
> * An [accessibility monitoring methodology](https://www.gov.uk/guidance/accessibility-monitoring-how-we-test) section (ie. the testing work).

Agree - I propose we start this as a result of this initial testing phase.

> * Maybe a note about open source. We'd have to be sure not to let this sound like an excuse, but rather a note about response times and that there's no support for this work.

💯  - this is completely acceptable and helps set clear expectations

> * A way for people to report they'd like to be reached out if/when we are testing (like [Gatsby's Surveys and Research section](https://www.gatsbyjs.com/accessibility-statement/#surveys-and-research-projects)). Since I think this could come off wrong, we may want to mention like "for outreach in any potential paid testing opportunities." At the same time, it'd be useful to know we have people available if we want to apply for funding this kind of work in the future.

+1 on this - do we want to use an email for this? if so which would be best?

> * [ ]  I filled this out for JupyterLab. Is there any supporting project we think should be included (ex. the notebook file)?
I think @gabalafou and @tonyfast  would have better ideas but right now I can only think of retrolab and by inheritance from JLab: binder and JupyterHub

> * [ ]  Where should this file go?
In the `/docs` directory - also we might want to ask folks in the projects above if they want to x-link to this from their own docs/repo for maximum coverage and avoid having 100 different statements in multiple repos

gabalafou commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+JupyterLab is designed to be compatible with the following technologies:
+
+* Windows, macOS, iOS, Android; Firefox, Chrome, Safari, Chromium browsers (mobile and desktop).
```
Oh wow, I've never opened JupyterLab on my smartphone. I assumed it wouldn't work...

gabalafou commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+An evaluation for JupyterLab is available at: [jupyterlab/jupyterlab/issues/9399](https://github.com/jupyterlab/jupyterlab/issues/9399).
+
+### Evaluation report
```
Two headings with same title?

isabela-pf replied:
Yep, that's what the generator made. 🤷

gabalafou commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+User reports on JupyterLab's accessibility are available at:[the jupyterlab/jupyterlab label `tag:accessibility`](https://github.com/jupyterlab/jupyterlab/labels/tag%3AAccessibility).
+
+### Formal complaints
```
Merge with Feedback? Feedback and Formal Complaints?

isabela-pf replied:
I think they are separate because formal complaints seems to be private and/or legal action related complaints, but we don't have a more formal channel at the moment so I'll put them together.

gabalafou commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+### Formal complaints
+
+* [Write an issue on jupyter/accessibility](https://github.com/jupyter/accessibility/issues/new)
+* [Write an issue on jupyterlab/jupyterlab](https://github.com/jupyterlab/jupyterlab/labels/tag%3AAccessibility) and request it be labeled tag:accessibility.
```
As an end user, how should I know which repo to submit my issue to?

isabela-pf replied:
I don't know, because I don't know what the difference would be in opening it on one repo or the other.

My estimate is that

Writing on jupyterlab/jupyterlab
- Would likely be the place where the change would need to be made (or in a related repo like Lumino)
- Would be the place where we may be able to get support from people not as focused on accessibility but knowledgeable about JupyterLab
- Has lots and lots of issues and could risk burying new issues in short amounts of time
- To be honest, I don't follow all JupyterLab issues by default because there's too much activity

Writing on jupyter/accessibility
- Would give a lot of attention to an issue
- Would likely require opening another issue on the repo where the code changes need to be made (at some point)
- Has less people actively following and may not have the fastest support

Does this make you feel more confident if we should advise only one? Or how we should tell people to choose?

gabalafou replied:
So I do think we should advise only one, to keep things simple. But I too am not entirely sure which one we should use.

I think for me, a big part of it is about where the accessibility statement lives. If it's in the jupyterlab repo, then I think we should point them to create issues on the jupyterlab repo (with accessibility tag, just like you're doing). If the statement lives in the accessibility repo, then they should create issues on the accessibility repo. I dunno. I don't feel too strongly about that. 

But I do think it's pretty confusing to give somebody two options without even a hint about why they should choose one over the other. 

I think our own lack of razor sharp clarity on this bleeds through to the accessibility statement. What I mean is, we don't seem to have ourselves a clear criteria for when to submit to one or the other. A clear criteria, for example, would be something like: for all accessibility issues pertaining to JupyterLab create an issue on the JupyterLab repo; for all other Jupyter projects (JupyterHub, Lumino, etc.) submit an issue to the accessibility repo.

trallard replied:
I thought we had decided to keep the statement in this repo and x-reference to it from the Jupyterlab readme/docs?

Suggestion - even if these sections have been merged

```
* For accessibility-related matters within the broader Jupyter ecosystem or directly related with this statement - [Write an issue on jupyter/accessibility](https://github.com/jupyter/accessibility/issues/new)
* For accessibility issues related to JupyterLab only -[Write an issue on jupyterlab/jupyterlab](https://github.com/jupyterlab/jupyterlab/labels/tag%3AAccessibility) and request it be labeled tag:accessibility.
```

gabalafou reviewed (commented):
Correct me if I'm wrong but it seems to me that the tl;dr of an accessibility statement for JupyterLab should be something like:

1. It sucks (as in, JupyterLab has a lot of egregious accessibility barriers)
2. But we care (we're actively working on it; here's a link to our roadmap)
3. And here's how to contact us (if you have any questions or concerns or want to get involved)

Do we feel that this accessibility statement conveys the essence of what we want to convey?

gabalafou commented:
> What seems to be inaccurate or missing from this current draft?
> What's on your wish list of statement additions?

A link to the roadmap?

> I filled this out for JupyterLab. Is there any supporting project we think should be included (ex. the notebook file)?

Because of the grant and my own mental well being, I have had to keep my focus a bit narrowly on JupyterLab. For me that includes paying attention to anything upstream (Lumino, CodeMirror) but giving carefully limited attention to anything downstream (Notebook v7). 

> Where should this file go?

An accessibility statement for JupyterLab should go in the JupyterLab repo, no?

Also I'm not as concerned about where the file lives as much as how to get to it. For example, I think if somebody visits the JupyterLab repo, they should be able to search the keyword "accessibility" and find a link to this statement on that page.

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+An evaluation for JupyterLab is available at: [jupyterlab/jupyterlab/issues/9399](https://github.com/jupyterlab/jupyterlab/issues/9399).
+
+### Evaluation report
```
```suggestion
```

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+## Edited from the [W3C accessibility statement generator](https://www.w3.org/WAI/planning/statements/generator/#create)
+
+
+## ⚠️ Draft Accessibility Statement for JupyterLab
```
```suggestion
## :construction: Draft Accessibility Statement for JupyterLab
```
I have realised that the other emoji renders widely different across OS and editors and is a nightmare

isabela-pf replied:
I also know the `:emoji:` format doesn't render everywhere, but we can do this if it's preferred.

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+## ⚠️ Draft Accessibility Statement for JupyterLab
+
+This is an accessibility statement from <span class="basic-information organization-name">Jupyter accessibility contributors</span>.
```
There seems to be a few `<span>` and the such straggling - do they need removing?

trallard commented on statement.md:
```diff
@@ -0,0 +1,84 @@
+
+Despite our best efforts to ensure accessibility of JupyterLab, there may be some limitations. Below is a description of known limitations, and potential solutions. Please contact us if you observe an issue not listed below.
+
+Known limitations for JupyterLab:
```
```suggestion
**Known limitations for JupyterLab:**
```

isabela-pf committed:
Apply suggestions from @trallard

Co-authored-by: Tania Allard <taniar.allard@gmail.com>

isabela-pf committed:
Apply suggestions from @trallard and @gabalafou

isabela-pf committed:
Apply suggestions from @trallard and @gabalafou

isabela-pf committed:
Move statement to docs/resources

isabela-pf commented:
Replying to non-inline comments!

@trallard 
- [x] I added a note about open-source response times under `### Feedback and Formal complaints`
- [x] I don't think we have any kind of private group contact point for testing or for the wider accessibility contributors, so if it makes sense to use my Quansight email we could use that. Edit: I've added a section to `###Feedback and Formal complaints`, but I left questions on it in inline feedback. Edit 2: I've added a contact.
- [x] I can add a note about JupyterLab being a foundation for other projects and that our accessibility is at least somewhat interlinked. I'm would rather not call out projects I don't think we've tested here, though I'm open if you have different thoughts. Edit: Added to `### Conformance Status`
- [x] I moved the file to `docs/resources`

@gabalafou 
- [x] I agree with you that that's a good structure for the statement. Do I think this does that? Eh… it does it very formally. I'm not confident it's approachable. I chose this mainly because it helped me get something done and gave me pointers to cover things I never would've though of in my inexperience. 

We could group the headers differently, though, and still benefit from W3C's guidance. Like use the outline you mentioned as a base for three main sections and lower the hierarchy of the existing headers to be grouped under those, if you think that would help navigation. Edit: I added headers roughly like you described and reordered the sections according to that.
- [x] I added a link to the roadmap under `### Measures to support accessibility`. Thanks for the reminder!
- [ ] An accessibility statement for JupyterLab should probably have a presence in the JupyterLab repo, yeah. Probably in the docs and in the repo itself somewhere. Do you think it should only be there, or does it belong in both JupyerLab and jupyter/accessibility? I don't think I have a strong opinion on that.

trallard commented on statement.md:
```diff
@@ -0,0 +1,89 @@
+## Edited from the [W3C accessibility statement generator](https://www.w3.org/WAI/planning/statements/generator/#create)
+
+
+## :construction: Draft Accessibility Statement for JupyterLab
```
Thanks for the reminder here @isabela-pf - I always forget there are a million md subsets - replacing with the emoji
```suggestion
## 🚧 Draft Accessibility Statement for JupyterLab
```

trallard commented on statement.md:
```diff
@@ -0,0 +1,89 @@
+
+### Date
+
+This statement was created on 16 May 2022 using the [W3C Accessibility Statement Generator Tool](https://www.w3.org/WAI/planning/statements/) with additions and edits from the Jupyter accessibility contributors community.
```
```suggestion
This statement was created on 16 May 2022 using the [W3C Accessibility Statement Generator Tool](https://www.w3.org/WAI/planning/statements/) with additions and edits from the Jupyter accessibility contributors community.

```
end of file line

trallard commented:
> I don't think we have any kind of private group contact point for testing or for the wider accessibility contributors, so if it makes sense to use my Quansight email we could use that.

I would rather just redirect folks to jupyter/accessibility instead of personal emails

> I can add a note about JupyterLab being a foundation for other projects and that our accessibility is at least somewhat interlinked. I'm would rather not call out projects I don't think we've tested here, though I'm open if you have different thoughts.

👍🏽  this sounds like a good approach

Thanks for all the work here @isabela-pf

isabela-pf committed:
Apply suggestions from @trallard

Co-authored-by: Tania Allard <taniar.allard@gmail.com>

isabela-pf committed:
Add research and intertwined ecossytem notes

isabela-pf commented on statement.md:
```diff
@@ -0,0 +1,92 @@
+
+* [Write an issue on jupyter/accessibility](https://github.com/jupyter/accessibility/issues/new)
+* [Write an issue on jupyterlab/jupyterlab](https://github.com/jupyterlab/jupyterlab/issues/new) and request it be labeled [tag:accessibility](https://github.com/jupyterlab/jupyterlab/labels/tag%3AAccessibility)
+* If you are interested in being part of any potential research or organized feedback initiatives, please ___. Gauging community interest in user research participation helps us gain the support to make it happen.
```
@trallard mentioned pointing people to the repo, but I think we need something a little more specific. I'm also unsure if a totally public place is the best since we may want to collect contact info. This is up for discussion.

I'm also happy to remove this section and more forward without it if it becomes too messy of a discussion.

```suggestion
* If you are interested in being part of any potential research or organized feedback initiatives, please [comment on this issue](). Gauging community interest in user research participation helps us gain the support to make it happen.
```
(Also oops to the accidental formatting I triggered. I'll fix that when we agree on what to do next.)

trallard replied:
I completely forgot - yes for now we can use your email and we should be able to request a Jupyter one too 🤔

trallard commented:
I think this is almost ready for merge? correct me if wrong @isabela-pf

isabela-pf committed:
Add feedback contact and reorder based on @gabalafou's feedback

isabela-pf commented:
@trallard I personally think this is clean enough to be merged. There's still some [feedback I haven't addressed](https://github.com/Quansight-Labs/accessibility/pull/3#issuecomment-1140059585) which is why I hadn't pushed for it. The core of the draft statement is here, though.

isabela-pf committed:
Move statement to existing file location

trallard commented:
The question on adding this to the JLab repo will be discussed in the upcoming a11y meeting. 
We can then proceed with either opening a PR with the statement or a PR linking to the statement here.

Apart from this - the PR looks good to merge, WDYT @isabela-pf ?

isabela-pf commented:
I am okay merging this! I think there is a solid enough foundation to build upon here.

isabela-pf commented on docs/resources/statement.md:
```diff
@@ -0,0 +1,106 @@
+
+JupyterLab is designed to be compatible with the following technologies:
+
+* Windows, macOS, iOS, Android; Firefox, Chrome, Safari, Chromium browsers (mobile and desktop).
```
@krassowski said add Linux!

isabela-pf commented on docs/resources/statement.md:
```diff
@@ -0,0 +1,106 @@
+JupyterLab is not compatible with:
+
+* Internet Explorer, Edge; JAWS, NVDA, VoiceOver, Narrator, Orca screen readers; voice control technology.
+
```
@krassowski said
- We are looking to support Edge
- Is it worth mentioning the things we want to support

isabela-pf commented on docs/resources/statement.md:
```diff
@@ -0,0 +1,106 @@
+
+JupyterLab's accessibility does not exist in isolation. JupyterLab inherits much from the many projects it is built upon and its accessibility conformance may impact projects built off of JupyterLab or its components. The accessibility of this ecosystem is interlinked, so conformance may need to be resolved at different levels in order to impact JupyterLab positively.
+
+#### Compatibility with browsers and assistive technology
```
@tonyfast says this might be better as a table! Sorting tech by os, browser, assistive tech, so on…

isabela-pf commented on docs/resources/statement.md:
```diff
@@ -0,0 +1,106 @@
+
+Please note that JupyterLab is an open-source project and that Jupyter accessibility contributors are a group defined on a voluntary basis. Like many other open-source projects, we cannot guarantee how long it may take to respond to and resolve an issue, though we do make an effort to do it as quickly as is possible with our resources.
+
+### Date
```
@tonyfast says this should be at the top. Note it's a living document.

isabela-pf committed:
Update document based on meeting feedback
